server/src/main/java/org/elasticsearch/cluster/coordination/PreVoteCollectorFactory.java

```diff
@@ -0,0 +1,204 @@
+        return maxTermSeen.accumulateAndGet(term, Math::max);
+    }
+
+    PreVoteCollectorFactory(Settings settings, PreVoteResponse preVoteResponse, TransportService transportService, LongConsumer startElection) {
+        super(settings);
+        this.preVoteResponse = preVoteResponse;
+        this.transportService = transportService;
+        this.startElection = startElection;
+
+        transportService.registerRequestHandler(REQUEST_PRE_VOTE_ACTION_NAME, Names.GENERIC, false, false,
+            PreVoteRequest::new,
+            (request, channel, task) -> channel.sendResponse(handlePreVoteRequest(request)));
+    }
+
+    public Releasable start(final ClusterState clusterState, final Iterable<DiscoveryNode> broadcastNodes) {
+        logger.debug("{} starting", this);
+        PreVotingRound currentRound = new PreVotingRound(clusterState);
+        currentRound.start(broadcastNodes);
+        return currentRound;
+    }
+
+    public void update(PreVoteResponse preVoteResponse, DiscoveryNode currentLeader) {
+        logger.trace("updating with preVoteResponse={}, currentLeader={}", preVoteResponse, currentLeader);
+        this.preVoteResponse = preVoteResponse;
+        this.currentLeader = currentLeader;
+    }
+
+    // package-private for testing
+    PreVoteResponse handlePreVoteRequest(PreVoteRequest request) {
+        updateMaxTermSeen(request.getCurrentTerm());
+        // TODO if we are a leader and the max term seen exceeds our term then we need to bump our term
+
+        DiscoveryNode currentLeader = this.currentLeader;
+        if (currentLeader == null || currentLeader.equals(request.getSourceNode())) {
+            return preVoteResponse;
+        } else {
+            throw new CoordinationStateRejectedException("rejecting " + request + " as there is already a leader");
+        }
+    }
+
+    // package-private for testing, actual usages TODO
+    long getMaxTermSeen() {
+        return maxTermSeen.get();
+    }
+
+    @Override
+    public String toString() {
+        return "PreVoteCollectorFactory{" +
+            "currentLeader=" + currentLeader +
+            ", preVoteResponse=" + preVoteResponse +
+            '}';
+    }
+
+    private class PreVotingRound implements Releasable {
+        private final Set<DiscoveryNode> preVotesReceived = newConcurrentSet();
+        private final AtomicBoolean electionStarted = new AtomicBoolean();
+        private final PreVoteRequest preVoteRequest;
+        private final ClusterState clusterState;
+        private final AtomicBoolean isRunning = new AtomicBoolean();
+
+        PreVotingRound(ClusterState clusterState) {
+            this.clusterState = clusterState;
+            preVoteRequest = new PreVoteRequest(transportService.getLocalNode(), preVoteResponse.getCurrentTerm());
+        }
+
+        void start(final Iterable<DiscoveryNode> broadcastNodes) {
+
+            final boolean isRunningChanged = isRunning.compareAndSet(false, true);
```

Comment:
this becomes obsolete by changing to isClosed

Reply:
Ok